File: src/lib/pipeline/supabase-pusher.ts
/**
 * supabase-pusher.ts — Push merged data to Supabase.
 *
 * Port of Python's src/supabase_pusher.py.
 * Handles upserting agents, daily metrics (with partial-update support),
 * quote records, quote duplicates, and upload history.
 *
 * CALL DATA HANDLING:
 * Agents can appear on both RingCentral (rc) and Ricochet (rico_ap/rico_ch).
 * When sources are uploaded separately, call data is ADDED (not overwritten).
 * A hidden JSONB column `call_source_breakdown` tracks each source's contribution
 * so re-uploads of the same source correctly replace (not accumulate).
 */

import type { SupabaseClient } from "@supabase/supabase-js"
import { SOURCE_FIELD_MAP, type QuoteRecord, type QuoteDuplicate } from "./types"

interface MergedRow {
  agent: string
  office: string
  team: string
  Calls: number
  Inbound: number
  Outbound: number
  TalkTimeSeconds: number
  Texts: number
  OutTexts: number
  OptIns: number
  OptOuts: number
  Quotes: number
  QuotesDeduped: number
  NB: number
  Items: number
  WrittenPremium: number
  NBAutoCount: number
  NBAutoItems: number
  PremPremium: number
  PremItems: number
  PremPoints: number
  callSourceBreakdown?: Record<string, { calls: number; inbound: number; outbound: number; talk_time_seconds: number }>
}

/** The four call-related fields that can come from multiple sources */
const CALL_FIELDS = new Set(["calls", "inbound", "outbound", "talk_time_seconds"])

/** Map from daily_metrics column name → callSourceBreakdown sub-key */
const CALL_FIELD_TO_BREAKDOWN_KEY: Record<string, string> = {
  calls: "calls",
  inbound: "inbound",
  outbound: "outbound",
  talk_time_seconds: "talk_time_seconds",
}

/**
 * Push merged data to Supabase.
 * Mirrors Python's push_to_supabase() exactly.
 */
export async function pushToSupabase(
  supabase: SupabaseClient,
  mergedData: MergedRow[],
  reportDate: string,
  uploadTypes: string[] | null,
  quoteRecords?: QuoteRecord[],
  quoteDuplicates?: QuoteDuplicate[],
  uploadId?: string | null,
  sourceAgents?: Record<string, Set<string>>,
): Promise<string[]> {
  const logs: string[] = []
  const isPartial = uploadTypes !== null && uploadTypes.length > 0
  const modeLabel = isPartial ? `PARTIAL (${uploadTypes!.join(",")})` : "FULL"
  logs.push(`[Supabase] Pushing data to Supabase (${modeLabel})...`)

  // Determine which fields to write in partial mode
  const partialFields = new Set<string>()
  if (isPartial) {
    for (const src of uploadTypes!) {
      const fields = SOURCE_FIELD_MAP[src] || []
      for (const f of fields) partialFields.add(f)
    }
    logs.push(`[Supabase] Partial update fields: ${[...partialFields].sort().join(", ")}`)
  }

  // 1. Upsert Agents
  const agentsData = mergedData.map(row => ({
    name: row.agent,
    team: row.team || "",
    office: row.office || "",
    active: true,
  }))

  if (agentsData.length > 0) {
    const { error } = await supabase
      .from("agents")
      .upsert(agentsData, { onConflict: "name" })

    if (error) {
      logs.push(`[Supabase] Failed to push agents: ${error.message}`)
      return logs
    }
  }

  // 2. Fetch Agent IDs
  const { data: agentsDB, error: agentFetchErr } = await supabase
    .from("agents")
    .select("id, name")

  if (agentFetchErr || !agentsDB) {
    logs.push(`[Supabase] Failed to fetch agents: ${agentFetchErr?.message}`)
    return logs
  }

  const agentIdMap = new Map<string, string>()
  for (const a of agentsDB) {
    agentIdMap.set(a.name, a.id)
  }

  // 3. Fetch existing daily metrics to preserve data
  let existingRows = new Map<string, Record<string, unknown>>()
  const existingManualData = new Map<string, Record<string, unknown>>()

  if (isPartial) {
    const { data: existingData } = await supabase
      .from("daily_metrics")
      .select("*")
      .eq("report_date", reportDate)

    if (existingData) {
      for (const row of existingData) {
        existingRows.set(row.agent_id, row)
        existingManualData.set(row.agent_id, {
          dismissed_todos: row.dismissed_todos,
          past_due_todos: row.past_due_todos,
          pivots: row.pivots,
        })
      }
    }
  } else {
    const { data: existingData } = await supabase
      .from("daily_metrics")
      .select("agent_id, dismissed_todos, past_due_todos, pivots, call_source_breakdown")
      .eq("report_date", reportDate)

    if (existingData) {
      for (const row of existingData) {
        existingManualData.set(row.agent_id, {
          dismissed_todos: row.dismissed_todos,
          past_due_todos: row.past_due_todos,
          pivots: row.pivots,
        })
        // For full mode, also store existing rows so we can read call_source_breakdown
        existingRows.set(row.agent_id, row)
      }
    }
  }

  // 4. Upsert Daily Metrics
  const metricsData: Record<string, unknown>[] = []

  for (const row of mergedData) {
    const agentId = agentIdMap.get(row.agent)
    if (!agentId) continue

    const existing = existingManualData.get(agentId) || {}
    const existingRow = existingRows.get(agentId) || {}

    // Build the full field map from pipeline data (non-call fields)
    const allFields: Record<string, unknown> = {
      calls: row.Calls,
      inbound: row.Inbound,
      outbound: row.Outbound,
      talk_time_seconds: row.TalkTimeSeconds,
      texts: row.Texts,
      out_texts: row.OutTexts,
      opt_ins: row.OptIns,
      opt_outs: row.OptOuts,
      quotes: row.Quotes,
      quotes_deduped: row.QuotesDeduped,
      nb_count: row.NB,
      items: row.Items,
      written_premium: row.WrittenPremium,
      nb_auto_count: row.NBAutoCount,
      nb_auto_items: row.NBAutoItems,
      prem_premium: row.PremPremium,
      prem_items: row.PremItems,
      prem_points: row.PremPoints,
      dismissed_todos: existing.dismissed_todos ?? 0,
      past_due_todos: existing.past_due_todos ?? 0,
      pivots: existing.pivots ?? 0,
    }

    // ---------- CALL SOURCE BREAKDOWN LOGIC ----------
    // Read existing breakdown from DB (if any)
    const existingBreakdown: Record<string, Record<string, number>> =
      (existingRow as Record<string, unknown>).call_source_breakdown as Record<string, Record<string, number>> || {}

    // New breakdown from the current merge
    const newBreakdown = row.callSourceBreakdown || {}

    // Compute the final merged breakdown:
    // - Sources in the current upload get REPLACED with new values
    // - Sources NOT in the current upload keep their existing values
    const finalBreakdown: Record<string, Record<string, number>> = { ...existingBreakdown }

    if (isPartial) {
      // Partial upload: merge new source contributions into existing breakdown
      // Only replace sources that are actually in this upload
      for (const [src, data] of Object.entries(newBreakdown)) {
        finalBreakdown[src] = { ...data }
      }
    } else {
      // Full upload: the merge already computed the complete breakdown
      // Replace the entire breakdown with what merge produced
      for (const key of Object.keys(finalBreakdown)) {
        delete finalBreakdown[key]
      }
      for (const [src, data] of Object.entries(newBreakdown)) {
        finalBreakdown[src] = { ...data }
      }
    }

    // Recompute call totals from the merged breakdown
    let totalCalls = 0, totalInbound = 0, totalOutbound = 0, totalTalkTime = 0
    for (const srcData of Object.values(finalBreakdown)) {
      totalCalls += srcData.calls || 0
      totalInbound += srcData.inbound || 0
      totalOutbound += srcData.outbound || 0
      totalTalkTime += srcData.talk_time_seconds || 0
    }

    // Override call fields with the correctly summed totals
    allFields.calls = totalCalls
    allFields.inbound = totalInbound
    allFields.outbound = totalOutbound
    allFields.talk_time_seconds = totalTalkTime

    if (isPartial) {
      // Start from existing row, only overwrite targeted fields
      const metric: Record<string, unknown> = {
        agent_id: agentId,
        report_date: reportDate,
      }

      for (const [field, newVal] of Object.entries(allFields)) {
        if (partialFields.has(field)) {
          // For call fields: use the recomputed total from breakdown
          // For non-call fields: use the value from allFields directly
          metric[field] = newVal
        } else {
          // Keep existing value, fall back to 0/default if no existing row
          const ev = (existingRow as Record<string, unknown>)[field]
          metric[field] = (ev !== null && ev !== undefined) ? ev : newVal
        }
      }

      // Always write the merged breakdown
      metric.call_source_breakdown = finalBreakdown

      metricsData.push(metric)
    } else {
      // Full mode — write everything
      metricsData.push({
        agent_id: agentId,
        report_date: reportDate,
        ...allFields,
        call_source_breakdown: finalBreakdown,
      })
    }
  }

  if (metricsData.length > 0) {
    const { error } = await supabase
      .from("daily_metrics")
      .upsert(metricsData, { onConflict: "agent_id,report_date" })

    if (error) {
      logs.push(`[Supabase] Failed to push metrics: ${error.message}`)
      return logs
    }
  }

  // 5. Push quote duplicates if provided
  if (quoteDuplicates && quoteDuplicates.length > 0) {
    logs.push(`[Supabase] Uploading ${quoteDuplicates.length} duplicate quote records...`)

    // Clear existing duplicates for these months
    const monthsToClear = [...new Set(quoteDuplicates.map(r => r.report_month))].sort()
    for (const rm of monthsToClear) {
      await supabase
        .from("quote_duplicates")
        .delete()
        .eq("report_month", rm)
    }

    // Insert in batches of 50
    let batchSuccess = 0
    for (let i = 0; i < quoteDuplicates.length; i += 50) {
      const batch = quoteDuplicates.slice(i, i + 50)
      const { error } = await supabase.from("quote_duplicates").insert(batch)
      if (!error) {
        batchSuccess += batch.length
      } else {
        logs.push(`[Supabase] ERROR batch ${i}: ${error.message}`)
      }
    }
    logs.push(`[Supabase] quote_duplicates: ${batchSuccess}/${quoteDuplicates.length} inserted.`)
  }

  // 6. Push individual quote records if provided
  if (quoteRecords && quoteRecords.length > 0) {
    logs.push(`[Supabase] Uploading ${quoteRecords.length} individual quote records...`)

    const quotePayloads: Record<string, unknown>[] = []
    for (const r of quoteRecords) {
      const aid = agentIdMap.get(r.agent)
      if (!aid) continue
      quotePayloads.push({
        agent_id: aid,
        report_date: r.report_date,
        quote_control_number: r.quote_control_number,
        product: r.product,
        premium: r.premium,
        sub_producer: r.sub_producer,
        upload_id: uploadId || null,
      })
    }

    if (quotePayloads.length > 0) {
      // Clear existing quote records for these dates
      const datesToClear = [...new Set(quotePayloads.map(r => r.report_date as string))].sort()
      logs.push(`[Supabase] Clearing existing quote records for ${datesToClear.length} date(s)`)
      for (const d of datesToClear) {
        await supabase.from("quote_records").delete().eq("report_date", d)
      }

      // Insert in batches of 100
      let batchSuccess = 0
      for (let i = 0; i < quotePayloads.length; i += 100) {
        const batch = quotePayloads.slice(i, i + 100)
        const { error } = await supabase.from("quote_records").insert(batch)
        if (!error) {
          batchSuccess += batch.length
        } else {
          logs.push(`[Supabase] ERROR quote_records batch ${i}: ${error.message}`)
        }
      }
      logs.push(`[Supabase] quote_records: ${batchSuccess}/${quotePayloads.length} inserted.`)
    }
  }

  // 7. Update upload history if we have an upload ID
  if (uploadId) {
    try {
      await supabase
        .from("upload_history")
        .update({
          status: "success",
          logs: logs.join("\n").substring(0, 2000),
        })
        .eq("id", uploadId)
    } catch {
      logs.push("[Supabase] Failed to update upload history")
    }
  }

  logs.push(`[Supabase] Successfully pushed ${metricsData.length} daily metrics for ${reportDate} (${modeLabel}).`)
  return logs
}

File: automation/src/supabase_pusher.py
import os
import json
import requests
import pandas as pd
from datetime import date

# Which Supabase columns each source type populates
SOURCE_FIELD_MAP = {
    "rc":          ["calls", "inbound", "outbound", "talk_time_seconds"],
    "rico_ap":     ["calls", "inbound", "outbound"],
    "rico_ch":     ["talk_time_seconds"],
    "hs":          ["texts", "out_texts", "opt_ins", "opt_outs"],
    "quotes":      ["quotes", "quotes_deduped"],
    "nb":          ["nb_count", "items", "written_premium", "nb_auto_count", "nb_auto_items"],
    "premium":     ["prem_premium", "prem_items", "prem_points"],
    "rico_leads":  [],  # writes to leads_snapshot table, not daily_metrics columns
}

def push_to_supabase(merged_df: pd.DataFrame, report_date: date, config: dict,
                     upload_types: list[str] | None = None,
                     actual_sources: list[str] | None = None,
                     quote_duplicates: list[dict] | None = None,
                     quote_records: list[dict] | None = None,
                     upload_id: str | None = None):
    """
    Pushes the merged DSR dataframe to Supabase via its REST API.
    Handles upserting agents and daily metrics.
    
    Parameters
    ----------
    upload_types : list[str], optional
        If provided (upload mode), only update the fields associated with
        these source types. Existing data for other fields is preserved.
    """
    supabase_url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL") or config.get("supabase", {}).get("url")
    supabase_key = os.environ.get("NEXT_PUBLIC_SUPABASE_ANON_KEY") or config.get("supabase", {}).get("key")

    if not supabase_url or not supabase_key:
        print("  [Supabase] Skip push: SUPABASE_URL or SUPABASE_ANON_KEY not found in config/env.")
        return

    is_partial = upload_types is not None and len(upload_types) > 0
    mode_label = f"PARTIAL ({','.join(upload_types)})" if is_partial else "FULL"
    print(f"  [Supabase] Pushing data to Supabase ({mode_label})...")
    
    headers = {
        "apikey": supabase_key,
        "Authorization": f"Bearer {supabase_key}",
        "Content-Type": "application/json",
        "Prefer": "resolution=merge-duplicates"
    }

    # Determine which fields to write in partial mode
    partial_fields = set()
    if is_partial:
        for src in upload_types:
            partial_fields.update(SOURCE_FIELD_MAP.get(src, []))
        print(f"  [Supabase] Partial update fields: {sorted(partial_fields)}")

    # 1. Upsert Agents
    agents_data = []
    for _, row in merged_df.iterrows():
        agent_payload = {
            "name": row.get("agent"),
            "team": row.get("team", ""),
            "office": row.get("office", ""),
            "active": True
        }
        for k, v in agent_payload.items():
            if pd.isna(v):
                agent_payload[k] = ""
        agents_data.append(agent_payload)

    if agents_data:
        res = requests.post(
            f"{supabase_url}/rest/v1/agents?on_conflict=name",
            headers=headers,
            json=agents_data
        )
        if res.status_code >= 400:
            print(f"  [Supabase] Failed to push agents: {res.text}")
            return

    # 2. Fetch Agent IDs
    res = requests.get(
        f"{supabase_url}/rest/v1/agents?select=id,name",
        headers=headers
    )
    if res.status_code >= 400:
        print(f"  [Supabase] Failed to fetch agents: {res.text}")
        return
        
    agents_db = res.json()
    agent_id_map = {a["name"]: a["id"] for a in agents_db}

    # 3. Fetch existing daily metrics to preserve data
    date_str = report_date.isoformat()
    if is_partial:
        # In partial mode, fetch full rows so we can merge
        res = requests.get(
            f"{supabase_url}/rest/v1/daily_metrics?report_date=eq.{date_str}&select=*",
            headers=headers
        )
    else:
        res = requests.get(
            f"{supabase_url}/rest/v1/daily_metrics?report_date=eq.{date_str}&select=agent_id,dismissed_todos,past_due_todos,pivots",
            headers=headers
        )
    
    existing_rows = {}
    existing_manual_data = {}
    if res.status_code < 400:
        for row in res.json():
            if is_partial:
                existing_rows[row["agent_id"]] = row
            existing_manual_data[row["agent_id"]] = {
                "dismissed_todos": row.get("dismissed_todos"),
                "past_due_todos": row.get("past_due_todos"),
                "pivots": row.get("pivots")
            }

    # 4. Upsert Daily Metrics
    metrics_data = []
    
    for _, row in merged_df.iterrows():
        name = row.get("agent")
        agent_id = agent_id_map.get(name)
        if not agent_id:
            continue
            
        def get_val(col, default=0, is_float=False):
            val = row.get(col, default)
            if pd.isna(val):
                return default
            if is_float:
                return float(val)
            return int(float(val))

        existing = existing_manual_data.get(agent_id, {})
        
        def preserve_manual(key):
            if existing.get(key) is not None:
                return existing[key]
            return get_val(key)

        # Build the full field map from pipeline data
        all_fields = {
            "calls": get_val("Calls"),
            "inbound": get_val("Inbound"),
            "outbound": get_val("Outbound"),
            "talk_time_seconds": get_val("TalkTimeSeconds"),
            "texts": get_val("Texts"),
            "out_texts": get_val("OutTexts"),
            "opt_ins": get_val("OptIns"),
            "opt_outs": get_val("OptOuts"),
            "quotes": get_val("Quotes"),
            "quotes_deduped": get_val("QuotesDeduped"),
            "nb_count": get_val("NB"),
            "items": get_val("Items"),
            "written_premium": get_val("WrittenPremium", 0.0, True),
            "nb_auto_count": get_val("NBAutoCount"),
            "nb_auto_items": get_val("NBAutoItems"),
            "prem_premium": get_val("PremPremium", 0.0, True),
            "prem_items": get_val("PremItems"),
            "prem_points": get_val("PremPoints", 0.0, True),
            "dismissed_todos": preserve_manual("dismissed_todos"),
            "past_due_todos": preserve_manual("past_due_todos"),
            "pivots": preserve_manual("pivots"),
        }

        if is_partial:
            # Start from existing row, only overwrite targeted fields
            existing_row = existing_rows.get(agent_id, {})
            metric = {
                "agent_id": agent_id,
                "report_date": date_str,
            }
            for field, new_val in all_fields.items():
                if field in partial_fields:
                    existing_val = existing_row.get(field)
                    # Only overwrite if: new data is non-zero, OR there's no existing data
                    if new_val and new_val != 0:
                        metric[field] = new_val
                    elif existing_val is not None:
                        metric[field] = existing_val
                    else:
                        metric[field] = new_val
                else:
                    # Keep existing value, fall back to 0/default if no existing row
                    metric[field] = existing_row.get(field) if existing_row.get(field) is not None else new_val
            metrics_data.append(metric)
        else:
            # Full mode — write everything
            metric = {"agent_id": agent_id, "report_date": date_str}
            metric.update(all_fields)
            metrics_data.append(metric)

    if metrics_data:
        res = requests.post(
            f"{supabase_url}/rest/v1/daily_metrics?on_conflict=agent_id,report_date",
            headers=headers,
            json=metrics_data
        )
        if res.status_code >= 400:
            print(f"  [Supabase] Failed to push metrics: {res.text}")
            return

    # 5. Upsert Leads Snapshot
    has_leads = not is_partial or (upload_types and "rico_leads" in upload_types)
    if has_leads:
        leads_data = []
        for _, row in merged_df.iterrows():
            name = row.get("agent")
            agent_id = agent_id_map.get(name)
            if not agent_id:
                continue
                
            def get_val(col, default=0):
                val = row.get(col, default)
                if pd.isna(val):
                    return default
                return int(float(val))
                
            leads_data.append({
                "agent_id": agent_id,
                "report_date": date_str,
                "contact": get_val("contact"),
                "quoted": get_val("quoted"),
                "hot": get_val("hot"),
                "xsale": get_val("xsale")
            })

        if leads_data:
            res = requests.post(
                f"{supabase_url}/rest/v1/leads_snapshot?on_conflict=agent_id,report_date",
                headers=headers,
                json=leads_data
            )
            if res.status_code >= 400:
                print(f"  [Supabase] Failed to push leads: {res.text}")
                return

    # 6. Upsert daily_reports_meta to mark eAgent as synced
    sources_to_check = actual_sources or upload_types or []
    if "screenshots" in sources_to_check or "eagent" in sources_to_check:
        try:
            from datetime import datetime as dt
            now_iso = dt.utcnow().isoformat() + "Z"
            meta_payload = {
                "report_date": date_str,
                "eagent_submitted": True,
                "submitted_at": now_iso,
                "updated_at": now_iso
            }
            res = requests.post(
                f"{supabase_url}/rest/v1/daily_reports_meta?on_conflict=report_date",
                headers=headers,
                json=meta_payload
            )
            if res.status_code >= 400:
                print(f"  [Supabase] Failed to update daily_reports_meta: {res.text}")
            else:
                print(f"  [Supabase] Marked eAgent as submitted in daily_reports_meta for {date_str}.")
        except Exception as e:
            print(f"  [Supabase] Error updating daily_reports_meta: {e}")

    # 7. Push quote duplicates if provided
    if quote_duplicates:
        print(f"  [Supabase] Uploading {len(quote_duplicates)} duplicate quote records...")
        months_to_clear = sorted(set(r["report_month"] for r in quote_duplicates))
        for rm in months_to_clear:
            requests.delete(
                f"{supabase_url}/rest/v1/quote_duplicates?report_month=eq.{rm}",
                headers={"apikey": supabase_key, "Authorization": f"Bearer {supabase_key}"}
            )
        
        batch_success = 0
        for i in range(0, len(quote_duplicates), 50):
            batch = quote_duplicates[i:i+50]
            resp = requests.post(
                f"{supabase_url}/rest/v1/quote_duplicates", 
                json=batch, 
                headers=headers
            )
            if resp.status_code in (200, 201, 204):
                batch_success += len(batch)
            else:
                print(f"  [Supabase] ERROR batch {i}: {resp.status_code} {resp.text}")
        print(f"  [Supabase] quote_duplicates: {batch_success}/{len(quote_duplicates)} inserted.")

    # 8. Push individual quote records if provided
    if quote_records:
        print(f"  [Supabase] Uploading {len(quote_records)} individual quote records...")
        quote_payloads = []
        for r in quote_records:
            aid = agent_id_map.get(r["agent"])
            if not aid:
                continue
            quote_payloads.append({
                "agent_id": aid,
                "report_date": r["report_date"],
                "quote_control_number": r["quote_control_number"],
                "product": r["product"],
                "premium": r["premium"],
                "sub_producer": r["sub_producer"],
                "upload_id": upload_id
            })

        if quote_payloads:
            # Delete existing quote records for the dates present in this batch to prevent duplicates
            dates_to_clear = sorted(set(r["report_date"] for r in quote_payloads))
            print(f"  [Supabase] Clearing existing quote records for {len(dates_to_clear)} date(s): {', '.join(dates_to_clear)}")
            for d in dates_to_clear:
                requests.delete(
                    f"{supabase_url}/rest/v1/quote_records?report_date=eq.{d}",
                    headers={"apikey": supabase_key, "Authorization": f"Bearer {supabase_key}"}
                )

            # Insert new quote records in batches
            batch_success = 0
            for i in range(0, len(quote_payloads), 100):
                batch = quote_payloads[i:i+100]
                resp = requests.post(
                    f"{supabase_url}/rest/v1/quote_records",
                    json=batch,
                    headers=headers
                )
                if resp.status_code in (200, 201, 204):
                    batch_success += len(batch)
                else:
                    print(f"  [Supabase] ERROR quote_records batch {i}: {resp.status_code} {resp.text}")
            print(f"  [Supabase] quote_records: {batch_success}/{len(quote_payloads)} inserted.")

    print(f"  [Supabase] Successfully pushed {len(metrics_data)} daily metrics for {date_str} ({mode_label}).")
